File: android/app/src/main/java/com/megadisplay/client/protocol/Handshake.kt
package com.megadisplay.client.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets

object Handshake {

    fun build(version: Int): ByteArray {
        val versionStr = String.format("%03d", version)
        val msg = (Protocol.HANDSHAKE_BASE + versionStr).toByteArray(StandardCharsets.US_ASCII) + 0.toByte()
        check(msg.size == 18) { "handshake must be 18 bytes, got ${msg.size}" }
        return msg
    }

    fun parse(data: ByteArray): Int {
        require(data.size >= 18) { "handshake too short: ${data.size}" }

        val first4 = String(data, 0, 4, StandardCharsets.US_ASCII)
        if (first4 == "MIRR") throw HandshakeException(HandshakeError.OLD_HOST)

        val base = String(data, 0, Protocol.HANDSHAKE_BASE.length, StandardCharsets.US_ASCII)
        if (base != Protocol.HANDSHAKE_BASE) throw HandshakeException(HandshakeError.WRONG_PROTOCOL)

        val versionStr = String(
            data,
            Protocol.HANDSHAKE_BASE.length,
            Protocol.HANDSHAKE_VERSION_LENGTH,
            StandardCharsets.US_ASCII,
        )
        val version = versionStr.toIntOrNull()
            ?: throw HandshakeException(HandshakeError.WRONG_PROTOCOL)

        require(data[Protocol.HANDSHAKE_BASE.length + Protocol.HANDSHAKE_VERSION_LENGTH] == 0.toByte()) {
            "missing NUL terminator"
        }

        val maxSupported = Protocol.SUPPORTED_HOST_VERSIONS[0]
        if (version > maxSupported) throw HandshakeException(HandshakeError.OLD_CLIENT)
        if (version < maxSupported && version !in Protocol.SUPPORTED_HOST_VERSIONS) {
            throw HandshakeException(HandshakeError.OLD_HOST)
        }

        return version
    }

    fun readHandshake(channel: java.nio.channels.ScatteringByteChannel): Int {
        val buf = ByteBuffer.allocate(18)

        buf.limit(4)
        while (buf.position() < 4) {
            val n = channel.read(buf)
            if (n < 0) throw java.io.EOFException("channel closed during handshake")
        }
        buf.flip()
        val first4 = ByteArray(4)
        buf.get(first4)
        if (String(first4, StandardCharsets.US_ASCII) == "MIRR") {
            throw HandshakeException(HandshakeError.OLD_HOST)
        }

        buf.clear()
        buf.put(first4)
        buf.limit(18)
        while (buf.position() < 18) {
            val n = channel.read(buf)
            if (n < 0) throw java.io.EOFException("channel closed during handshake")
        }
        buf.flip()

        return parse(ByteArray(18).also { buf.get(it) })
    }
}

enum class HandshakeError { WRONG_PROTOCOL, OLD_HOST, OLD_CLIENT }

class HandshakeException(val error: HandshakeError) : Exception(error.name)

File: driver/megadisplay-protocol/src/handshake.rs
//! Protocol handshake constants and logic.

pub const HANDSHAKE_BASE: &str = "KELOCUBE_MIRR_";
pub const HANDSHAKE_VERSION_LENGTH: usize = 3;
pub const SUPPORTED_HOST_VERSIONS: &[u32] = &[4, 3];

/// The handshake message sent by both sides: `"KELOCUBE_MIRR_" + 3-digit version + NUL`.
///
/// Total length: 18 bytes (14 base + 3 version + 1 NUL).
pub fn build_handshake(version: u32) -> Vec<u8> {
    let version_str = format!("{version:03}");
    let mut msg = HANDSHAKE_BASE.as_bytes().to_vec();
    msg.extend(version_str.as_bytes());
    msg.push(0);
    assert_eq!(msg.len(), 18);
    msg
}

/// Parse a received handshake, returning the peer's version.
///
/// Returns `Err` if the handshake is invalid or the version is unsupported.
pub fn parse_handshake(data: &[u8]) -> Result<u32, HandshakeError> {
    if data.len() < 18 {
        return Err(HandshakeError::TooShort);
    }

    // Legacy V1 check: first 4 bytes "MIRR" → reject
    if &data[..4] == b"MIRR" {
        return Err(HandshakeError::OldHostVersion);
    }

    // Check base string
    if &data[..HANDSHAKE_BASE.len()] != HANDSHAKE_BASE.as_bytes() {
        return Err(HandshakeError::WrongProtocol);
    }

    // Parse version
    let version_str = std::str::from_utf8(
        &data[HANDSHAKE_BASE.len()..HANDSHAKE_BASE.len() + HANDSHAKE_VERSION_LENGTH],
    )
    .map_err(|_| HandshakeError::WrongProtocol)?;
    let version: u32 = version_str
        .parse()
        .map_err(|_| HandshakeError::WrongProtocol)?;

    // Check NUL terminator
    if data[HANDSHAKE_BASE.len() + HANDSHAKE_VERSION_LENGTH] != 0 {
        return Err(HandshakeError::WrongProtocol);
    }

    // Version validation
    let max_supported = SUPPORTED_HOST_VERSIONS[0];
    if version > max_supported {
        return Err(HandshakeError::OldClientVersion);
    }
    if version < max_supported && !SUPPORTED_HOST_VERSIONS.contains(&version) {
        return Err(HandshakeError::OldHostVersion);
    }

    Ok(version)
}

#[derive(Debug, Clone, thiserror::Error)]
pub enum HandshakeError {
    #[error("handshake too short")]
    TooShort,
    #[error("wrong protocol")]
    WrongProtocol,
    #[error("client version is too old for this host")]
    OldClientVersion,
    #[error("host version is too old")]
    OldHostVersion,
}
